### PR TITLE
ftp: adds a config option ftp-hash for autofp-scheduler

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/5205

Describe changes:
- Adds an option `ftp-hash` for `autofp-scheduler` : like `hash` except for FTP-ish flows

No S-V test as this is about a concurrency issue...

Modifies #7165 with rebase and commit description including ticket number

### CODE REVIEW

**Comment by victorjulien** [comment]
I understand what is done here, but I feel it may be a bit too specific. Could we just do ippair+ipproto and be done with it?

**Comment by catenacyber** [comment]
> I understand what is done here, but I feel it may be a bit too specific. Could we just do ippair+ipproto and be done with it?

I am not sure...
I guess this may offer better load balancing than ippair+ipproto...
There is also the related question wether this new value should be the default (to have consistency over performance)

**Comment by suricata-qa** [comment]
Information:
 

ERROR: QA failed on tlpw1_files_sha256.
 | field | test | baseline | % |
|--|--|--|--| 
| | tlpr1_stats_chk | | |
| .flow.memuse | 531156288 | 498197568 | 106.62% | 

Pipeline 7243

File: src/flow-hash.c [hunk 1]
```diff
@@ -107,6 +107,61 @@ typedef struct FlowHashKey6_ {
     };
 } FlowHashKey6;
 
+uint32_t FlowGetFtpHash(const Packet *p)
```
Comment (victorjulien) [inline src/flow-hash.c:110]: this is not specific to ftp, right? Its essentially ippair+ipproto? If so lets reflect that in the name

File: src/tmqh-flow.c [hunk 2]
```diff
@@ -272,6 +277,33 @@ void TmqhOutputFlowIPPair(ThreadVars *tv, Packet *p)
...
 }
 
+static void TmqhOutputFlowFTPHash(ThreadVars *tv, Packet *p)
+{
+    uint32_t qid;
+    TmqhFlowCtx *ctx = (TmqhFlowCtx *)tv->outctx;
+
+    if (p->flags & PKT_WANTS_FLOW) {
+        uint32_t hash = p->flow_hash;
+        if (p->tcph != NULL && ((p->sp > 1024 && p->dp > 1024) || p->dp == 21 || p->sp == 21)) {
```
Comment (victorjulien) [inline src/tmqh-flow.c:287]: should this be `>= 1024`?

also, should this include port 20?

File: suricata.yaml.in [hunk 3]
```diff
@@ -1139,6 +1139,8 @@ host-mode: auto
 #
 # hash     - Flow assigned to threads using the 5-7 tuple hash.
 # ippair   - Flow assigned to threads using addresses only.
+# ftp-hash - Flow assigned to threads using the hash, except for FTP, so that
+# ftp-data flows will be handled by the same thread
```
Comment (victorjulien) [inline suricata.yaml.in:1143]: looks like this line wrap needs better alignment

Comment (catenacyber) [reply suricata.yaml.in:1143]: indeed

File: src/tmqh-flow.c (hunk 2 above)
Comment (catenacyber) [reply src/tmqh-flow.c:287]: > should this be `>= 1024`?

Right

> also, should this include port 20?

Not sure : if it is port 20, it is ftp-data without expectation, right ?

File: src/flow-hash.c (hunk 1 above)
Comment (catenacyber) [reply src/flow-hash.c:110]: You are right (named it because it was ftp-friendly)

**Comment by catenacyber** [comment]
Replaced by https://github.com/OISF/suricata/pull/7353

File: src/tmqh-flow.c (hunk 2 above)
Comment (victorjulien) [reply src/tmqh-flow.c:287]: No, active ftp mode also uses expectations.

Comment (victorjulien) [reply src/tmqh-flow.c:287]: 2 SV tests using port 20: `bug-4877` and `filestore-v2.7-stream-depth`

**Pull request closed without merging**